File: checkout.rb
#!/usr/local/bin/ruby

require 'store'
require 'cart'

store = Store.new
cart = Cart.new(store)

print "Enter skus (name quantity price), one per line, enter blank line to continue\n"

while gets
  line = $_.chomp
  break if line =~ /^$/
  if line =~ /^(\w+)\s+(\d+)\s+(\d*\.?\d{0,2})$/
    name      = $1
    quantity  = $2.to_i
    price     = $3.to_f
  else
    next
  end
  store.add_product(name, quantity, price)
end

print store

print "Enter products to purchase, one per line, enter blank line to continue\n"
while gets
  line = $_.chomp
  break if line =~ /^$/
  next unless store.products.keys.include?(line)
  cart.add_item(line)
end

print cart
printf("TOTAL $%.2f\n", cart.total)

File: checkout.pl
#!/usr/bin/perl

use strict;
use Data::Dumper;

my %products;
my %cart;

print "Enter products (name, quantity, price), empty line to continue...\n";
while(<STDIN>) {
	chomp;
	last if /^$/;
	if(/^(\w+)\s+(\d+)\s+(\d*\.?\d{0,2})$/) {
		my $name = $1;
		my $quantity = $2;
		my $price = $3;
		$products{$name}{$quantity} = $price;
	}
}

print "Enter your order, empty line to continue\n";
while(<STDIN>) {
	chomp;
	last if /^$/;
	if(exists $products{$_}) {
		$cart{$_}++;
	}
}

my $total = 0;
foreach my $product (sort keys %cart) {
	my $quantity = $cart{$product};
	my $price = 0;
	while($quantity > 0) {
		foreach my $discount (reverse sort keys %{$products{$product}}) {
			my $mod = $quantity % $discount;
			my $div = sprintf("%d", $quantity / $discount);
			if($mod >= 0) {
				$price = $price + ($div * $products{$product}{$discount});
				$quantity = $quantity - ($div * $discount);
			} else {
				$price = $price + $products{$product}{$discount};
				$quantity = $quantity - $discount;
			}
		}
	}

	printf("%s %2d \$%.2f\n", $product, $cart{$product}, $price);
	$total += $price;
}
printf("TOTAL \$%.2f\n", $total);

